### ISSUE TITLE
[Paper]Contrastive Multi-modal Knowledge Graph Representation Learning

### ISSUE THREAD

**Issue by Vannora9** [Vannora9/KG#4]
# Information

- Year: 2022
- Conference: IEEE
- Author: Fang, Quan, Xiaowei Zhang, Jun Hu, Xian Wu, and Changsheng Xu.
- Organization: Institute of Automation, Chinese Academy of Sciences.

# General
_本文提出一种对比性多模态知识图谱表示学习的方法，对于一个知识节点，同时包含文字信息和图片信息，而且在图谱中的关系也包含不同的关系类型，例如文中提到的Obama节点， _In addition, the left blue subgraph shows his political role as a former president of the USA, while the right orange one shows his family role as a husband and a father, which displays multi-view semantics._

![image](https://user-images.githubusercontent.com/86482592/235828945-6c6d8b73-3ae7-4d52-a344-75d247e70f1f.png)

本文提出Contrastive Multi-modal Graph Neural Network (CMGNN)，overall如下图所示。

![image](https://user-images.githubusercontent.com/86482592/235829532-225d82d0-bdc2-458b-9542-8aa4ed72b57e.png)

本方法主要分为三个部分：Multi-modal Knowledge Graph， Local Contrastive Multi-relational Feature Aggregation 和High-order
Contrastive Encoding。该模型首先使用多层感知机网络为不同模态的数据学习实体嵌入。其次，局部对比多关系特征聚合模块将实体的局部邻域关系类型区分为多个子图。第三，设计了一个关于实体及其子图的内部和外部关系特征聚合网络，以获取特定关系的实体特征。然后，我们使用关系特定的注意力来聚合来自不同衍生子图的消息，以获得统一的实体特征。接下来，对实体及其高阶邻居子图进行高阶对比编码模块，以生成实体嵌入。最后，编码器将实体嵌入输出到解码器，该解码器是一个KGRL模型，可以替换为许多现有的KGRL模型。该模型选择了ConvE作为解码器。

### Multi-modal Knowledge Graph
![image](https://user-images.githubusercontent.com/86482592/235830250-9c5cc97c-780a-4695-afad-31a69257ff5f.png)
其中，输入是multi-modal knowledge graph，嵌入层包括**结构实体建模（structural entity modeling）和多模态编码（multi-modal encoding）**。

1. 对于构建实体建模，使用one-hot实体ID向量，将实体ID通过一个带有权重矩阵Wid的嵌入层传递，生成一个稠密向量e。具体来说，Wid是一个矩阵，它的每一行对应一个实体ID的嵌入向量，矩阵的列数是嵌入向量的维度。当我们将一个实体ID传递到这个嵌入层时，就相当于在Wid中找到这个实体ID对应的行，然后取出这一行作为嵌入向量e。
2. 多模态编码主要考虑文本和图像。
    1. 文本：使用Word2Vec进行词向量训练，然后应用平滑反向频率(SIF)模型来获取一个句子的加权平均值，用作表示文本特征的句子向量。
    2. 使用在ImageNet上预训练的ResNet50的最后一个隐藏层。
        - 隐藏层：将文本和图像的特征向量串联起来，通过权重矩阵`$W^{mul}$`转化为紧凑向量e’，然后将e和e’传入多层感知集中。其定义如下：
        
![image](https://user-images.githubusercontent.com/86482592/235832516-4bddc5a6-5a0b-4319-bff9-b1f0adadf992.png)
最后通过该隐藏层获得了输入实体的抽象表示 $h_i^{(n)} \in R^d$

### Local Contrastive Multi-relational Feature Aggregation
用图卷积网络对相同实体进行聚合。对于每个实体和其关系特定的子图，进行对比学习模块，以捕获复杂的结构信息。通过局部多关系邻居聚合的操作，捕获并存储了每个实体编码中的局部接近度结构。其中包含**内部关系特征聚合（Intra-relational Feature Aggregation Module）和外部关系特征聚合（Inter-Relational Feature Aggregation Module）**

1. 内部关系特征聚合：提取与所关注实体具有相同关系类型的邻居，得到该实体的特定关系子图（即一阶邻居子图）。实体特征更新定义如下：
![image](https://user-images.githubusercontent.com/86482592/235835386-d5d13288-310f-4588-9ec7-4c8acc51ac19.png)
2. 外部关系特征聚合：采用均值池化操作处理每个子图，并获得每个关系子图的向量表示，通过跨关系特征聚合，自动学习不同关系的重要性，并将它们组合生成最终表示。提出基于关系特定的注意力机制，首先通过非线性转换来转换嵌入，然后使用一个共享的注意力向量 $q ∈ R^d$ 来计算注意力值。
![image](https://user-images.githubusercontent.com/86482592/235835727-25a671cf-ba31-4531-8caf-30ed805f78fa.png)
attention values with the softmax function to get the weigh
![image](https://user-images.githubusercontent.com/86482592/235835779-e4f8edbc-2fbc-457b-a653-781433efb868.png)
3. 对比约束：为了有效地从局部的多关系三元组中提取全面的特征。对初始节点特征进行全局洗牌和添加高斯噪声来生成负样本，这样可以帮助模型更好地学习数据的结构和特征，从而提高模型的性能。（图二中的KG Diffusion）。the local contrastive objective is defined as
![image](https://user-images.githubusercontent.com/86482592/235836242-bfaa1218-5d62-4b34-b4bb-969b4a680ed1.png)

### High-Order Contrastive Encoding
1. 采用重要性传播采样程序提取关注实体ei的高阶邻居图。
    1. 沿着知识图谱链对关注实体e迭代得到多个h-hop路径
    2. 聚合所有路径并计算出现频率来得到每个节点的重要性分数
    3. 按分数降序排列后选择一组排名靠前的实体，被选中实体以及其连接被定义为e的高阶邻居图。
高阶对比约束目标定义为
![image](https://user-images.githubusercontent.com/86482592/235837005-18cd61cd-f90a-4be4-bf99-d2ff3a2120f4.png)

## 解码
使用ConvE作为解码器。将实体和关系嵌入重塑为二维张量，然后对其进行卷积，以建立实体和关系之间的相互作用。卷积层的输出然后通过全连接层，以产生一个给定的三元组的最终得分。score function of ConvE：
![image](https://user-images.githubusercontent.com/86482592/235837428-456e6eb1-9e97-44f0-a4cf-299c35862ee9.png)

loss function of the encoder
![image](https://user-images.githubusercontent.com/86482592/235837457-9ba529c0-afe9-44d8-a3f0-0ee22fc29152.png)
Combining the KG embedding and contrastive constraint, we have the following overall objective function:
![image](https://user-images.githubusercontent.com/86482592/235837540-3cc02617-7c47-41af-9bb6-497b87bb2261.png)

#Evaluate
### dataset
![image](https://user-images.githubusercontent.com/86482592/235837735-5c0a9525-1647-4a74-8999-038c2028df87.png)

##Result
### Link Prediction
![image](https://user-images.githubusercontent.com/86482592/235837822-7de7bbea-dacf-4992-8b6b-d0c2e9405af0.png)
![image](https://user-images.githubusercontent.com/86482592/235837882-e1be3e6e-98dd-4bc0-af2e-c263c164f6a5.png)
![image](https://user-images.githubusercontent.com/86482592/235837902-97974472-772e-4f5b-9935-666aa3eef8c1.png)

### Entity Classification
![image](https://user-images.githubusercontent.com/86482592/235837977-24256a09-e2e4-40ed-a0ef-d17d44d6d116.png)
![image](https://user-images.githubusercontent.com/86482592/235838001-37f33461-59b1-42ac-ae56-5a0d418f0cbb.png)

# Some unfamiliar concepts
### 平滑反向频率(SIF)
通过加权平均词向量的方式来生成句子向量。它考虑了词频的影响，将常见词的权重降低，使得它们在句子表示中的重要性较小，而将不常见的词的权重提高，使得它们在句子表示中的重要性更高。
![image](https://user-images.githubusercontent.com/86482592/235831782-c49e3eaf-81c0-4bab-a979-12c3497a3570.png)
